src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java

```diff
@@ -0,0 +1,274 @@
+                        "Failed to get roles from JWT claims. Check if this key is correct and available in the JWT payload.");
+                roles = new String[0];
+            } else {
+                final String encryptedRoles = claims.get("roles").toString();
+                //TODO: WHERE TO GET THE ENCRYTION KEY
+                final String decryptedRoles = EncryptionDecryptionUtil.decrypt(encryptionKey, encryptedRoles);
+                roles = Arrays.stream(decryptedRoles.split(",")).map(String::trim).toArray(String[]::new);
+            }
+
+            if (subject == null) {
+                log.error("No subject found in JWT token");
+                return null;
+            }
+
+            if (audience == null) {
+                log.error("No audience found in JWT token");
+            }
+
+            final AuthCredentials ac = new AuthCredentials(subject, roles).markComplete();
+
+            for(Entry<String, Object> claim: claims.entrySet()) {
+                ac.addAttribute("attr.jwt."+claim.getKey(), String.valueOf(claim.getValue()));
+            }
+
+            return ac;
+
+        } catch (WeakKeyException e) {
+            log.error("Cannot authenticate user with JWT because of ", e);
+            return null;
+        } catch (Exception e) {
+            if(log.isDebugEnabled()) {
+                log.debug("Invalid or expired JWT token.", e);
+            }
+            return null;
+        }
+    }
+
+    @Override
+    public boolean reRequestAuthentication(final RestChannel channel, AuthCredentials creds) {
+        final BytesRestResponse wwwAuthenticateResponse = new BytesRestResponse(RestStatus.UNAUTHORIZED,"");
+        wwwAuthenticateResponse.addHeader("WWW-Authenticate", "Bearer realm=\"OpenSearch Security\"");
+        channel.sendResponse(wwwAuthenticateResponse);
+        return true;
+    }
+
+    @Override
+    public String getType() {
+        return "extension_jwt";
+    }
+
+    //TODO: Extract the audience (ext_id) and inject it into thread context
+
+    protected String extractSubject(final Claims claims, final RestRequest request) {
+        String subject = claims.getSubject();
+        if(subjectKey != null) {
+            // try to get roles from claims, first as Object to avoid having to catch the ExpectedTypeException
+            Object subjectObject = claims.get(subjectKey, Object.class);
+            if(subjectObject == null) {
+                log.warn("Failed to get subject from JWT claims, check if subject_key '{}' is correct.", subjectKey);
+                return null;
+            }
+            // We expect a String. If we find something else, convert to String but issue a warning
+            if(!(subjectObject instanceof String)) {
+                log.warn("Expected type String for roles in the JWT for subject_key {}, but value was '{}' ({}). Will convert this value to String.", subjectKey, subjectObject, subjectObject.getClass());
+            }
```

Comment:
In the original `HTTPJwtAuthenticator.java` it does the same thing. I believe the reason to do this is that we also need to handle the case for some nonString claims. In the test case, we also have a test case called `testNonStringClaim`. I think this scenario only happens when the user enable the BWCMode, which means instead of passing encoded roles, plain text roles will be passed into the claim. For the disabling of BMCMode, the role will always be encrypted, so that it will always be an encrypted `string`.

Reply:
The message here shouldn't contain the word `roles`. Can you update the message to `Expected type String for subject...`

This should be updated in the `HTTPJwtAuthenticator` as well. There is overlap between here and the `HTTPJwtAuthenticator` class. Maybe we can create a utility class with static methods or maybe add these to an abstract class that both inherit from?